File: VB/WindowCalculationExample/TabPageLoadHelper.vb
Imports DevExpress.DashboardCommon
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq

Namespace WindowCalculationExample

    Friend Class TabPageLoadHelper

        Private _TabContainerName As String, _ScatterChartItemName As String, _PivotItemName As String

        Private originalDashboard As Dashboard = New Dashboard()

        Private destinationParentContainer As TabContainerDashboardItem

        Public Property Dashboard As Dashboard

        Public Property TabContainerName As String
            Get
                Return _TabContainerName
            End Get

            Private Set(ByVal value As String)
                _TabContainerName = value
            End Set
        End Property

        Public Property ScatterChartItemName As String
            Get
                Return _ScatterChartItemName
            End Get

            Private Set(ByVal value As String)
                _ScatterChartItemName = value
            End Set
        End Property

        Public Property PivotItemName As String
            Get
                Return _PivotItemName
            End Get

            Private Set(ByVal value As String)
                _PivotItemName = value
            End Set
        End Property

        Public Sub New(ByVal xmldashboardFile As String)
            originalDashboard.LoadFromXml(Path.Combine(Environment.CurrentDirectory, xmldashboardFile))
        End Sub

        Public Sub PrepareDashboard()
            Dim targetDashboard As Dashboard = New Dashboard()
            destinationParentContainer = New TabContainerDashboardItem()
            targetDashboard.Items.Add(destinationParentContainer)
            TabContainerName = destinationParentContainer.ComponentName
            Dim tabPage As DashboardTabPage = CopyDashboardToTabPage(targetDashboard, originalDashboard)
            Dashboard = targetDashboard
            Dim tabPage1 As DashboardTabPage = destinationParentContainer.CreateTabPage()
            targetDashboard.RebuildLayout()
            CopyTabPage(tabPage, tabPage1)
            ScatterChartItemName = Enumerable.Single(targetDashboard.Items, Function(item)(TypeOf item Is ScatterChartDashboardItem) AndAlso item.ParentContainer Is tabPage1).ComponentName
            PivotItemName = Enumerable.Single(targetDashboard.Items, Function(item) TypeOf item Is PivotDashboardItem AndAlso item.ParentContainer Is tabPage1).ComponentName
        End Sub

        Private Function CopyDashboardToTabPage(ByVal targetDashboard As Dashboard, ByVal originalDashboard As Dashboard) As DashboardTabPage
            Dim tabPage As DashboardTabPage = destinationParentContainer.CreateTabPage()
            Dim layoutPage As DashboardLayoutTabPage = New DashboardLayoutTabPage(tabPage)
            Dim layoutTabContainer As DashboardLayoutTabContainer = New DashboardLayoutTabContainer(destinationParentContainer)
            layoutTabContainer.ChildNodes.Add(layoutPage)
            Dim itemsAndGroups As IEnumerable(Of DashboardItem) = originalDashboard.Items.Union(originalDashboard.Groups).Where(Function(item) Not(TypeOf item Is TabContainerDashboardItem))
            For Each item As DashboardItem In itemsAndGroups
                If item.ParentContainer Is Nothing Then
                    If Not(TypeOf item Is TabContainerDashboardItem) Then
                        targetDashboard.Items.Add(item)
                        item.ParentContainer = tabPage
                    End If
                End If
            Next

            Dim layoutRoot As DashboardLayoutGroup = originalDashboard.LayoutRoot
            layoutPage.ChildNodes.Add(layoutRoot)
            targetDashboard.LayoutRoot = New DashboardLayoutGroup()
            targetDashboard.LayoutRoot.ChildNodes.Add(layoutTabContainer)
            Return tabPage
        End Function

        Public Sub CopyTabPage(ByVal tabPageSource As DashboardTabPage, ByVal tabPageDestination As DashboardTabPage)
            Dim targetDashboard As Dashboard = Dashboard
            Dim tabPageSourceLayout As DashboardLayoutTabPage = tabPageSource.Dashboard.LayoutRoot.FindRecursive(tabPageSource)
            Dim tabPageDestinationLayout As DashboardLayoutTabPage = tabPageSource.Dashboard.LayoutRoot.FindRecursive(tabPageDestination)
            LayoutCopy(tabPageSourceLayout, tabPageDestinationLayout)
        End Sub

        Private Sub LayoutCopy(ByVal copyFrom As DashboardLayoutNode, ByVal copyTo As DashboardLayoutNode)
            copyTo.Weight = copyFrom.Weight
            If(TypeOf copyTo Is DashboardLayoutGroup) AndAlso (TypeOf copyFrom Is DashboardLayoutGroup) Then CType(copyTo, DashboardLayoutGroup).Orientation = CType(copyFrom, DashboardLayoutGroup).Orientation
            If TypeOf copyFrom Is DashboardLayoutGroup Then
                For Each currentNode As DashboardLayoutNode In CType(copyFrom, DashboardLayoutGroup).ChildNodes
                    Dim node As DashboardLayoutNode = New DashboardLayoutItem()
                    Dim item As DashboardItem = currentNode.DashboardItem
                    If item IsNot Nothing Then
                        node.DashboardItem = CopyItemFromPageToPage(item, destinationParentContainer.TabPages(1))
                        node.Weight = currentNode.Weight
                    End If

                    If TypeOf currentNode Is DashboardLayoutGroup Then
                        node = CopyTree(currentNode)
                    End If

                    CType(copyTo, DashboardLayoutGroup).ChildNodes.Add(node)
                Next
            End If
        End Sub

        Private Function CopyItemFromPageToPage(ByVal item As DashboardItem, ByVal itemContainer As IDashboardItemContainer) As DashboardItem
            Dim itemCopy As DashboardItem = item.CreateCopy()
            itemCopy.ComponentName = GetNameOnNextPage(item.ComponentName)
            itemCopy.Name = GetNameOnNextPage(item.Name)
            itemCopy.ParentContainer = itemContainer
            Dashboard.Items.Add(itemCopy)
            Return itemCopy
        End Function

        Private Function GetNameOnNextPage(ByVal name As String) As String
            Return name & "_1"
        End Function

        Private Function CopyTree(ByVal currentNode As DashboardLayoutNode) As DashboardLayoutNode
            Dim dashboardLayoutNode As DashboardLayoutGroup = New DashboardLayoutGroup()
            LayoutCopy(currentNode, dashboardLayoutNode)
            Return dashboardLayoutNode
        End Function
    End Class
End Namespace

File: CS/WindowCalculationExample/TabPageLoadHelper.cs
using DevExpress.DashboardCommon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace WindowCalculationExample
{
    class TabPageLoadHelper
    {
        private Dashboard originalDashboard = new Dashboard();
        private TabContainerDashboardItem destinationParentContainer;
        public Dashboard Dashboard { get; set; }
        public string TabContainerName { get; private set; }
        public string ScatterChartItemName { get; private set; }
        public string PivotItemName { get; private set; }

        public TabPageLoadHelper(string xmldashboardFile)
        {
            originalDashboard.LoadFromXml(Path.Combine(Environment.CurrentDirectory, xmldashboardFile));
        }

        public void PrepareDashboard()
        {
            Dashboard targetDashboard = new Dashboard();
            destinationParentContainer = new TabContainerDashboardItem();
            targetDashboard.Items.Add(destinationParentContainer);
            TabContainerName = destinationParentContainer.ComponentName;

            DashboardTabPage tabPage = CopyDashboardToTabPage(targetDashboard, originalDashboard);

            this.Dashboard = targetDashboard;
            DashboardTabPage tabPage1 = destinationParentContainer.CreateTabPage();
            targetDashboard.RebuildLayout();

            CopyTabPage(tabPage, tabPage1);

            ScatterChartItemName = targetDashboard.Items.Single(item => (item is ScatterChartDashboardItem) && (item.ParentContainer == tabPage1)).ComponentName;
            PivotItemName = targetDashboard.Items.Single(item => item is PivotDashboardItem && (item.ParentContainer == tabPage1)).ComponentName;

        }

        private DashboardTabPage CopyDashboardToTabPage(Dashboard targetDashboard, Dashboard originalDashboard)
        {
            DashboardTabPage tabPage = destinationParentContainer.CreateTabPage();
            DashboardLayoutTabPage layoutPage = new DashboardLayoutTabPage(tabPage);
            DashboardLayoutTabContainer layoutTabContainer = new DashboardLayoutTabContainer(destinationParentContainer);
            layoutTabContainer.ChildNodes.Add(layoutPage);

            IEnumerable<DashboardItem> itemsAndGroups = originalDashboard.Items.Union(originalDashboard.Groups).Where(item => !(item is TabContainerDashboardItem));
            foreach (DashboardItem item in itemsAndGroups)
            {
                if (item.ParentContainer == null)
                {
                    if (!(item is TabContainerDashboardItem))
                    {
                        targetDashboard.Items.Add(item);
                        item.ParentContainer = tabPage;
                    }
                }
            }

            DashboardLayoutGroup layoutRoot = originalDashboard.LayoutRoot;
            layoutPage.ChildNodes.Add(layoutRoot);
            targetDashboard.LayoutRoot = new DashboardLayoutGroup();
            targetDashboard.LayoutRoot.ChildNodes.Add(layoutTabContainer);
            return tabPage;
        }

        public void CopyTabPage(DashboardTabPage tabPageSource, DashboardTabPage tabPageDestination)
        {
            Dashboard targetDashboard = this.Dashboard;
            DashboardLayoutTabPage tabPageSourceLayout = tabPageSource.Dashboard.LayoutRoot.FindRecursive(tabPageSource);
            DashboardLayoutTabPage tabPageDestinationLayout = tabPageSource.Dashboard.LayoutRoot.FindRecursive(tabPageDestination);
            LayoutCopy(tabPageSourceLayout, tabPageDestinationLayout);
        }

        void LayoutCopy(DashboardLayoutNode copyFrom, DashboardLayoutNode copyTo)
        {
            copyTo.Weight = copyFrom.Weight;
            if ((copyTo is DashboardLayoutGroup) && (copyFrom is DashboardLayoutGroup))
                ((DashboardLayoutGroup)copyTo).Orientation = ((DashboardLayoutGroup)copyFrom).Orientation;
            if (copyFrom is DashboardLayoutGroup)
            {
                foreach (DashboardLayoutNode currentNode in ((DashboardLayoutGroup)copyFrom).ChildNodes)
                {
                    DashboardLayoutNode node = new DashboardLayoutItem();
                    DashboardItem item = currentNode.DashboardItem;
                    if (item != null) {
                        node.DashboardItem = CopyItemFromPageToPage(item, destinationParentContainer.TabPages[1]);
                        node.Weight = currentNode.Weight;
                    }
                    if (currentNode is DashboardLayoutGroup)
                    {
                        node = CopyTree(currentNode);
                    }
                    ((DashboardLayoutGroup)copyTo).ChildNodes.Add(node);
                }
            }
        }

        private DashboardItem CopyItemFromPageToPage(DashboardItem item, IDashboardItemContainer itemContainer)
        {
            DashboardItem itemCopy = item.CreateCopy();
            itemCopy.ComponentName = GetNameOnNextPage(item.ComponentName);
            itemCopy.Name = GetNameOnNextPage(item.Name);
            itemCopy.ParentContainer = itemContainer;
            Dashboard.Items.Add(itemCopy);
            return itemCopy;
        }

        private string GetNameOnNextPage(string name)
        {
            return name + "_1";
        }

        private DashboardLayoutNode CopyTree(DashboardLayoutNode currentNode)
        {
            DashboardLayoutGroup dashboardLayoutNode = new DashboardLayoutGroup();
            LayoutCopy(currentNode, dashboardLayoutNode);
            return dashboardLayoutNode;
        }
    }
}
